### PR TITLE
samples: drivers: adc: add demo and fix STM32 sequence configuration

### PR DESCRIPTION
**Pull request by martinjaeger**
This sample shows how to use the ADC API. If the ADC peripheral is found in the devicetree, the first two channels are read and printed.

While writing the sample I realized that the STM32 driver does not yet support a sequence of multiple channels. For most other MCUs this seems to be the case.

As the STM32 silently ignored all channels except for the first one of the sequence, I added a check so that it returns an error if more than one channel is requested. The driver should be updated to properly support sequences in a second step.

### CODE REVIEW

File: samples/drivers/adc/src/main.c
Comment (gmarull) [inline samples/drivers/adc/src/main.c:15]: Would probably be good to require an alias as some other samples do (e.g. blinky)

File: samples/drivers/adc/src/main.c
Comment (gmarull) [inline samples/drivers/adc/src/main.c:43]: If I'm not wrong not all ADC support the same settings (e.g. resolution may vary, reference...)

File: samples/drivers/adc/src/main.c
Comment (gmarull) [inline samples/drivers/adc/src/main.c:98]: It would probably be good to have an endless loop (sample every second?) so that user can play with the analog signal and see the actual value changes,

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:15]: Yeah, thought about that, but I was wondering if this sample is worth touching almost every board dts. But that would be the cleanest approach, so I agree.
Are there any preferences for the alias name? Just `adc`?

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:43]: Ok, will double-check.

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:98]: Agreed, makes sense.

**Comment by martinjaeger** [comment]
@gmarull I implemented the endless loop for reading the samples as you suggested.

Also started with adding aliases to each board that supports an ADC. But I realized that this would only simplify the `main.c` in the sample, but not ensure that the sample runs with all boards that have an ADC alias, as they might have different references, gains or resolutions. This information is not available from the Devicetree and ADCs are not as simple as GPIOs for LEDs, for example.

I checked how the ADC tests deal with those variants: They have about 200 lines of code at the beginning to cover the different ADC settings. I would rather not add this to the sample to ensure correct settings for every possible ADC. So I suggest to keep the approach as it is, which aims to support most common ADCs. And it compiles well for all 49 boards with `adc` enabled in their .yaml file.

**Comment by gmarull** [comment]
> @gmarull I implemented the endless loop for reading the samples as you suggested.
> 
> Also started with adding aliases to each board that supports an ADC. But I realized that this would only simplify the `main.c` in the sample, but not ensure that the sample runs with all boards that have an ADC alias, as they might have different references, gains or resolutions. This information is not available from the Devicetree and ADCs are not as simple as GPIOs for LEDs, for example.
> 
The current approach doesn't guarantee either that sample will run if `adc`/`adc0`/`adc1`... is defined, so I'd still prefer to have an alias for adc. It should only be added to boards that are known to work with the settings defined in the sample. Current approach may create false expectations, i.e. it works if any of these DT nodes are present.

**Comment by martinjaeger** [comment]
> The current approach doesn't guarantee either that sample will run if `adc`/`adc0`/`adc1`... is defined, so I'd still prefer to have an alias for adc. It should only be added to boards that are known to work with the settings defined in the sample. Current approach may create false expectations, i.e. it works if any of these DT nodes are present.

Just realized that it's possible to filter sanitycheck builds based on existence of an aliases. That would allow this approach and not make it overly complex. So I'll give it another go.

**Comment by martinjaeger** [comment]
@gmarull Now it's based on aliases, as you suggested. Please have a look again.

**Review by gmarull — APPROVED** [review]

**Comment by martinjaeger** [comment]
@erwango and @FRASTM can one of you please do a second review of this PR? Thanks.

File: samples/drivers/adc/src/main.c
Comment (ABOSTM) [inline samples/drivers/adc/src/main.c:96]:
```suggestion
		for (uint8_t ch = ADC_1ST_CHANNEL_ID; ch < ADC_1ST_CHANNEL_ID + ADC_NUM_CHANNELS;
```

File: samples/drivers/adc/src/main.c
Comment (ABOSTM) [inline samples/drivers/adc/src/main.c:98]:
```suggestion
			printk(" %d", m_sample_buffer[ch - ADC_1ST_CHANNEL_ID]);
```

File: samples/drivers/adc/src/main.c
Comment (ABOSTM) [inline samples/drivers/adc/src/main.c:78]: See my general comment.
Modification for ADC_1ST_CHANNEL_ID different from 0
```suggestion
	for (uint8_t ch = ADC_1ST_CHANNEL_ID; ch < ADC_1ST_CHANNEL_ID + ADC_NUM_CHANNELS; ch++) {
```

**Review by ABOSTM — DISMISSED** [review]
Hi @martinjaeger ,
Thanks for this PR.
I tried the demo with nucleo_l476rg, and I faced several issues:
1) in DT, the channel used for ADC is channel1, and the good reason to not use channel0 is that channel0 is connected to Vref, so it is not usable for normal adc measurement.
But in the sample, the 1st channel is hardcoded to 0
`#define ADC_1ST_CHANNEL_ID	0`
Should sample retrieve channel from DT ?

2) Once I modified the sample for test purpose
`#define ADC_1ST_CHANNEL_ID	1`
I still have issue because the sample is made to work only with 1st channel 0.
See inline comments, how I solved this.

**Comment by martinjaeger** [comment]
@ABOSTM Thanks a lot for your quick response. In fact I didn't realize that some Nucleo boards don't start with ADC1_IN0 wired to Arduino A0 pin, as this was the case for both boards I had available for testing (F072 and L073). It seems like all Nucleo boards have ADC1_IN1 wired to one of the A0-A5 pins though, so this seems like a better choice than channel 0.

In order to define the channel via devicetree we would have to use a custom binding with a phandle-array, as far as I understand. We could use the [voltage-divider](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/iio/adc/voltage-divider.yaml), but that would make the example more complex. Or do you have a different suggestion how to specify the channel via Devicetree?

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c]: Changed this to just loop from 0 to ADC_NUM_CHANNELS in order to make clear that the values in the buffer start from index 0 independent of the selected channels.

For the other loop I implemented the fix as you suggested. Thanks!

**Comment by ABOSTM** [comment]
@martinjaeger , I had this in mind https://github.com/zephyrproject-rtos/zephyr/pull/25996
but after new read-through, I am no so sure it fit the needs.

**Comment by martinjaeger** [comment]
> @martinjaeger , I had this in mind #25996
> but after new read-through, I am no so sure it fit the needs.

Yeah, that could be helpful in the future, so that we can find out the used ADC channels from pinmux devicetree... but I just realized that the previous fix to use channel 1 instead of 0 is not a clean solution either, as it does not match with the `pinmux.c` of all boards. Even though high impedance is the default pin state, it would be better to match the ADC channel with what is defined in the pinmux.

Seems like there is just no simple solution to demonstrate the ADC without too many #ifdefs to cover all the different cases. I will think about it again and remove the 2.4 label for now, so we have more time to get it right.

**Comment by martinjaeger** [comment]
As discussed with @mbolivar-nordic on Slack, I updated the sample to select the channel based on a devicetree overlay using the new `zephyr,user` node.

The provided (and tested) overlay for the Nucleo L073 should be easily adjustable for other MCUs. I also added an overlay for nRF52 DK with two channels (not supported by STM32 driver at the moment). As I don't have any Nordic board, it would be great if you could test this @mbolivar-nordic. If you suggest to use a different board, just let me know.

**Review by mbolivar-nordic — COMMENTED** [review]
> As I don't have any Nordic board, it would be great if you could test this @mbolivar-nordic. If you suggest to use a different board, just let me know.

I really appreciate you doing this; thank you.

It doesn't work, though, and it's making me question the way that we currently use io-channels for ADC in Zephyr.

The problem is that adc_nrfx_saadc.c selects `CONFIG_ADC_CONFIGURABLE_INPUTS`, and requires a nonzero `input_positive` field to work.

For example, this makes it work (just a hack to show what I mean, not actually asking for this change):

```diff
diff --git a/samples/drivers/adc/boards/nrf52dk_nrf52832.overlay b/samples/drivers/adc/boards/nrf52dk_nrf52832.overlay
index 13d48bb20d..3c3ab84c36 100644
--- a/samples/drivers/adc/boards/nrf52dk_nrf52832.overlay
+++ b/samples/drivers/adc/boards/nrf52dk_nrf52832.overlay
@@ -6,7 +6,7 @@
 
 / {
        zephyr,user {
-               /* Arduino header pins A0 and A1 */
-               io-channels = <&adc 1>, <&adc 2>;
+               /* any channel number <=7 is arbitrary and will work */
+               io-channels = <&adc 7>;
        };
 };
diff --git a/samples/drivers/adc/src/main.c b/samples/drivers/adc/src/main.c
index ad00b2d480..9c5e3fb4b7 100644
--- a/samples/drivers/adc/src/main.c
+++ b/samples/drivers/adc/src/main.c
@@ -51,13 +51,16 @@ static uint8_t channel_ids[ADC_NUM_CHANNELS] = {
 
 static int16_t sample_buffer[ADC_NUM_CHANNELS];
 
+#include <hal/nrf_saadc.h>
+
 struct adc_channel_cfg channel_cfg = {
        .gain = ADC_GAIN,
        .reference = ADC_REFERENCE,
        .acquisition_time = ADC_ACQUISITION_TIME,
        /* channel ID will be overwritten below */
        .channel_id = 0,
-       .differential = 0
+       .differential = 0,
+       .input_positive = NRF_SAADC_INPUT_AIN1, /* HACK */
 };
 ```

I think that this sample should work for ADC drivers that either do or do not select `CONFIG_ADC_CONFIGURABLE_INPUTS`; the question therefore is just how to make it work using device configuration in devicetree, since that's clearly where this belongs.

A quick and easy but unsatisfactory way would be to add more properties to `/zephyr,user` to specify the input fields in `struct adc_channel_cfg`. This doesn't feel right since `devicetree/iio.h` is supposed to be able to represent ADC channel configuration in devicetree, if I understand correctly.

I thought maybe we should do something like this in `nordic,nrf-saadc.yaml`:

```yaml
...
    "#io-channel-cells":
      const: 4

io-channel-cells:
    - channel
    - differential
    - input-positive
    - input-negative
```

But that seems wrong if we're trying to reuse iio style bindings, because the Linux binding says typically 0 or 1 suffices:

```
Required properties:
#io-channel-cells: Number of cells in an IIO specifier; Typically 0 for nodes
		   with a single IIO output and 1 for nodes with multiple
		   IIO outputs.
```

Poking around elsewhere in linux/Documentation/devicetree, I found [this example](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/iio/adc/adc.txt), which configures individual channels as sub-nodes of the ADC node:

```
        adc@0 {
                compatible = "some,adc";
                ...
                channel@0 {
                        bipolar;
                        diff-channels = <0 1>;
                        ...
                };

                channel@1 {
                        diff-channels = <2 3>;
                        ...
                };
        };
```

If drivers which require configurable inputs used this sort of child binding, I think it could be made to work. But our `adc-controller.yaml` has no such `child-binding:`, so this would be a change in the bindings as well as the relevant drivers.

I do think that's probably the right way to do this, though.

Curious what others (especially @galak, @brix, @anangl, @pabigot) think.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Putting a -1 on this for now until we've sorted out the configurable inputs issue in my above comment, and adding dev-review so it hopefully doesn't take too long to resolve.

**Comment by martinjaeger** [comment]
Great feedback @mbolivar-nordic. I agree that the approach using child-binding is probably the best solution. And I think that all other `adc_channel_cfg` parameters should also go into the child binding so that we don't need >100 lines of code as in this sample just to read one ADC channel.

I guess the necessary changes would be beyond the scope of this PR, though. So we might just remove the last commit for nRF support again and start a discussion about rework of ADC bindings in a dedicated issue?

**Comment by mbolivar-nordic** [comment]
@martinjaeger I think that my concern is a blocker for the sample too.

If we do it the way I suggest, the sample itself should no longer be performing the ADC channel initialization. The devicetree will contain a channel configuration in child nodes of the ADC node, and I think that means the ADC driver is responsible for ensuring boot-time configuration in that case. That would mean proper usage in a sample would just need to call adc_read(), not adc_channel_setup().

Note I am not saying adc_channel_setup() should be removed from the API -- it's still useful for runtime reconfiguration, for example. But it does seem to me that if we put boot time configuration for drivers that select ADC_CONFIGURABLE_INPUTS into DT, drivers that don't should behave the same way, for abstraction's sake.

**Comment by anangl** [comment]
> If we do it the way I suggest, the sample itself should no longer be performing the ADC channel initialization. The devicetree will contain a channel configuration in child nodes of the ADC node, and I think that means the ADC driver is responsible for ensuring boot-time configuration in that case. That would mean proper usage in a sample would just need to call adc_read(), not adc_channel_setup().

That sounds reasonable to me. A common function in `adc_context.h` could be provided for this boot-time configuration of channels, and particular drivers would just need to call it from their initialization function.

File: samples/drivers/adc/src/main.c
Comment (pabigot) [inline samples/drivers/adc/src/main.c:104]: There is API (`adc_raw_to_millivolts()`) that converts the raw reading to millivolts; it would be nice to include that in the sample, as many people will find that more useful than the raw value.

**Comment by martinjaeger** [comment]
@mbolivar-nordic Do you anticipate a quick & simple solution for the Nordic ADC drivers with the new devicetree features merged in #30406 or should I remove the last commit and we postpone Nordic support for the ADC sample?

This sample has been open for quite a while now and I would like to get it into 2.5 if possible so that it can already provide some guidance for users of other MCU series.

**Comment by mbolivar-nordic** [comment]
I don't anticipate a quick solution. Sorry, I thought I had communicated with you on Slack that I thought the best way to get this merged in the 2.5 timeframe was to put it into `samples/boards` since it's hardware specific.

**Comment by martinjaeger** [comment]
> ... I thought the best way to get this merged in the 2.5 timeframe was to put it into `samples/boards` since it's hardware specific.

Ok, must have misunderstood that.

If it's a requirement for a sample outside the `boards` folder that it works with **all** platforms, I'm fine with moving it into the `boards` folder, even though it's not limited to one board or even to one platform. I'm afraid that people might not find the sample there if they are looking for an ADC driver sample.

What about taking out the last commit and adding a comment about which platforms are currently supported by this sample? Should be at least STM32, NXP Kinetis, Microchip SAM0.

What do you think, @anangl?

**Comment by mbolivar-nordic** [comment]
> What about taking out the last commit and adding a comment about which platforms are currently supported by this sample? Should be at least STM32, NXP Kinetis, Microchip SAM0.

That seems OK tome

File: samples/drivers/adc/README.rst
Comment (mbolivar-nordic) [inline samples/drivers/adc/README.rst:19]:
```suggestion
Microchip SAM0. Nordic nRFx MCUs are currently not supported.
```

The stated  reason why is not totally accurate, so better to just say they're not supported IMO.

File: samples/drivers/adc/README.rst
Comment (mbolivar-nordic) [inline samples/drivers/adc/README.rst:27]:
```suggestion
In addition to that, this sample requires an ADC channel specified in the ``io-channels``
```

Otherwise users might think this is a general requirement of the API, I think.

File: samples/drivers/adc/src/main.c
Comment (mbolivar-nordic) [inline samples/drivers/adc/src/main.c:25]: Hmm, this isn't doing what you want it to. DT_PHANDLE_BY_IDX returns a node identifier, which [is not a value](https://docs.zephyrproject.org/latest/guides/dts/api-usage.html#node-identifiers-are-not-values).

I think what is likely happening is that comparing the two node identifiers like this is silently converted to `0 != 0` in the preprocessor. (I bet you would get a warning with `-Wundef`.)

If you really want to keep this check, maybe you can compare the `device` structure addresses intead?

File: samples/drivers/adc/src/main.c
Comment (mbolivar-nordic) [inline samples/drivers/adc/src/main.c:75]: I think it would be nicer to start using DEVICE_DT_GET in new samples. I think that style is easier to read:

```
const struct device *dev_adc = DEVICE_DT_GET(ADC_NODE);

if (!device_is_ready(dev_adc)) {
	printk("ADC device not found\n");
	return;
}
```

Then you don't need to define `ADC_DEVICE_NAME`.

This is not blocking, just a suggestion.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Almost there. Just a few minor things.

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:25]: Oops, good spot. I'm going to wrap it into `DT_REG_ADDR()` to make it comparable.

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:97]: Unfortunately, the API `adc_ref_internal()` is not supported by all drivers. Most STM32 MCUs for example use the voltage at an external VDDA pin (which is usually equal to the MCU supply voltage) as the reference and `adc_ref_internal()` always returns 0 (not set by the driver because the supply voltage is unknown).

Without knowing the ADC reference voltage, also the `adc_raw_to_millivolts` API cannot be used properly, so I suggest to keep the raw readings in the sample until handling of the reference voltage is improved.

**Comment by martinjaeger** [comment]
@mbolivar-nordic Thanks for your review. I addressed all your comments and I agree that it makes sense to use the new `DEVICE_DT_GET` macros.

Unfortunately, it fails to compile. I pushed it anyway so that you can also see the error in buildkite:
```
/workdir/zephyr/samples/drivers/adc/src/main.c: In function 'main':
/workdir/zephyr/include/device.h:46:40: error: '__device_DT_N_S_soc_S_adc_40012400' undeclared (first use in this function)
46 | #define DEVICE_NAME_GET(name) (_CONCAT(__device_, name))
   |                                        ^~~~~~~~~

[...]

/workdir/zephyr/samples/drivers/adc/src/main.c:68:33: note: in expansion of macro 'DEVICE_DT_GET'
 68 |  const struct device *dev_adc = DEVICE_DT_GET(ADC_NODE);
    |                                 ^~~~~~~~~~~~~
```

DT device macro support for the ADC has been merged in #30567, so I guess it should work. Would `__device_DT_N_S_soc_S_adc_40012400` normally also end up in `devicetree_unfixed.h`? I can't find it there or in any other file in the build folder. Am I doing something obviously wrong? (CC @pabigot)

File: samples/drivers/adc/src/main.c
Comment (pabigot) [reply samples/drivers/adc/src/main.c:97]: OK.  But you could fetch the reference voltage, and if it's not zero provide the millivolt conversion.  Then the sample would work as desired on boards that do support this feature, and it would be a clue to users for what they need to do to get the measurement in a useful unit.

**Comment by pabigot** [comment]
> DT device macro support for the ADC has been merged in #30567, so I guess it should work. Would `__device_DT_N_S_soc_S_adc_40012400` normally also end up in `devicetree_unfixed.h`? I can't find it there or in any other file in the build folder. Am I doing something obviously wrong? (CC @pabigot)

For now you need `DEVICE_DT_DECLARE(ADC_NODE);` at file scope to expose the symbol that is referenced by `DEVICE_DT_GET()`.  That will change when #31291 is merged.

**Comment by martinjaeger** [comment]
Alright, thanks for clarification @pabigot. Will rebase as soon as mentioned PR is merged. And I will also add the millivolts reading at the same time.

File: samples/drivers/adc/src/main.c
Comment (pabigot) [inline samples/drivers/adc/src/main.c:68]: adc_value must be `int32_t` as a pointer to it is passed to `adc_raw_to_millivolts` and that would be wrong if `int` was 64-bit.  (Probably isn't, but still.)

Generally we prefer to use types with size specified unless the API involved doesn't.  So both these values should be `int32_t`, while `err` should remain `int`.

But see proposed change below which moves these definitions.

File: samples/drivers/adc/src/main.c
Comment (pabigot) [inline samples/drivers/adc/src/main.c:99]: I think it's worth providing both the raw and the millivolt values where available.  This patch does it; refactor as you wish.

```
@@ -97,18 +100,22 @@ void main(void)
 
 		printk("ADC reading(s):");
 		for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
-			adc_value = sample_buffer[i];
+			int32_t raw_value = sample_buffer[i];
+			int32_t mv_value = INT32_MAX;
+
+			printk(" %d", raw_value);
 			if (adc_vref > 0) {
 				/*
 				 * Convert raw reading to millivolts if driver
 				 * supports reading of ADC reference voltage
 				 */
+				mv_value = raw_value;
 				adc_raw_to_millivolts(adc_vref, ADC_GAIN,
-					ADC_RESOLUTION, &adc_value);
+					ADC_RESOLUTION, &mv_value);
+				printk(" = %d mV", mv_value);
 			}
-			printk(" %d", adc_value);
 		}
-		printk(" (%s)\n", adc_vref > 0 ? "in mV" : "raw");
+		printk("\n");
 
 		k_sleep(K_MSEC(1000));
 	}
```

produces:
```
*** Booting Zephyr OS build zephyr-v2.4.0-3072-g36eaaa4e3b1c  ***               
ADC reading(s): 164 = 24 mV                                                     
ADC reading(s): 302 = 44 mV                                                     
ADC reading(s): 274 = 40 mV
```

File: samples/drivers/adc/src/main.c
Comment (pabigot) [inline samples/drivers/adc/src/main.c:80]: This patch makes this work on Nordic nRF52 devices:
```
@@ -78,6 +78,9 @@ void main(void)
 	 */
 	for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
 		channel_cfg.channel_id = channel_ids[i];
+#ifdef CONFIG_ADC_NRFX_SAADC
+		channel_cfg.input_positive = SAADC_CH_PSELP_PSELP_AnalogInput0 + channel_ids[i];
+#endif
 		adc_channel_setup(dev_adc, &channel_cfg);
 
 		sequence.channels |= BIT(channel_ids[i]);
```

**Review by pabigot — CHANGES_REQUESTED** [review]
Need to fix the argument type to `adc_raw_to_millivolts`, and would like to see both raw and millivolt outputs.

Nordic support is not necessary if we don't want to put ADC-specific code in the generic sample.  I'm fine with leaving it out, just needed it to test.

File: samples/drivers/adc/src/main.c
Comment (martinjaeger) [reply samples/drivers/adc/src/main.c:79]: Ok, added your suggestion together with an overlay for nrf52dk_nrf52832. Unfortunately can't test because I don't have any Nordic board.

**Review by pabigot — APPROVED** [review]

File: samples/drivers/adc/src/main.c
Comment (pabigot) [reply samples/drivers/adc/src/main.c:79]: It works for me:
```
*** Booting Zephyr OS build zephyr-v2.4.0-3073-g07bc2d3f55a5  ***               
ADC reading(s): 697 = 102 mV   771 = 112 mV                                     
ADC reading(s): 978 = 143 mV   4095 = 599 mV                                    
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): -415 = -61 mV   228 = 33 mV                                     
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): 4095 = 599 mV   4095 = 599 mV  
ADC reading(s): 1484 = 217 mV   1597 = 233 mV                                   
ADC reading(s): 992 = 145 mV   1121 = 164 mV   
```
(The max values are me putting my fingers on the AIN pins. Not sure what the negative is; perhaps the default configuration is double-ended.)

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
https://github.com/zephyrproject-rtos/zephyr/pull/25590#discussion_r558372732

I don't mean to be a pain, but I don't think we should merge this additional patch.

There are two separate concepts here: *channel* and *input pin*. In the SAADC, each channel can sample any input pin. This patch is mixing the two concepts in a way that doesn't make sense, in my opinion.

@martinjaeger if you want to compare two nodes for equality, I think using DT_DEP_ORD is a bit nicer than DT_REG_ADDR.

**Comment by pabigot** [comment]
> There are two separate concepts here: _channel_ and _input pin_. In the SAADC, each channel can sample any input pin. This patch is mixing the two concepts in a way that doesn't make sense, in my opinion.

I'm fine with dropping the Nordic support patch; I agree they're different concepts.  I'm unaware of any other ADC that has Nordic's flexibility, and that distinction probably shouldn't have been part of the general API.  It would be better to have Nordic-specific API to map channels to input sources, just as other vendors might do so with pinmux.

Perhaps a middle ground would be to instead add a comment in the sample documentation saying that the sample can't be made to work on Nordic and why,  with a reference to the Nordic battery sample for people who want to know how to use ADC on that platform:

> Note: This sample does not work on Nordic platforms where there is a distinction between channel and analog input that requires additional configuration.  See samples/boards/nrf/battery for an example of using the ADC infrastructure on Nordic hardware.

**Comment by martinjaeger** [comment]
@pabigot Removed the Nordic overlay again and added note as suggested.

@mbolivar-nordic Used new `DT_SAME_NODE` macro that was just merged for the DT node comparison.

Hope it's ready to go now :)

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I believe @ABOSTM 's comments have been addressed and this is ready to go for 2.5, but I will leave that decision up to the release team.

**Pull request merged**
